subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java

```diff
@@ -687,6 +689,34 @@ public List<SubscriptionBillingEvent> getSubscriptionBillingEvents(final Version
         }
     }
 
+
+    private DateTime alignToNextBCDIfRequired(final Plan curPlan, final PlanPhase curPlanPhase, final DateTime originalDate, final SubscriptionCatalog catalog, final Integer bcdLocal, final InternalTenantContext context) throws SubscriptionBaseApiException, CatalogApiException {
+
+        if (!apiService.isEffectiveDateForExistingSubscriptionsAlignedToBCD(context)) {
+            return originalDate;
+        }
+
+        final BillingAlignment billingAlignment = catalog.billingAlignment(new PlanPhaseSpecifier(curPlan.getName(), curPlanPhase.getPhaseType()),
+                                                                           originalDate, originalDate);
+
+        final int accountBillCycleDayLocal = apiService.getAccountBCD(context);
+        Integer bcd = bcdLocal;
+        if (bcd == null) {
+            bcd = BillCycleDayCalculator.calculateBcdForAlignment(null, this, this, billingAlignment, context, accountBillCycleDayLocal);
+        }
+
+        // TODO : no Recurring PlanPhase, Usage sections, ...
+        final BillingPeriod billingPeriod = curPlanPhase.getRecurring() != null ? curPlanPhase.getRecurring().getBillingPeriod() : BillingPeriod.NO_BILLING_PERIOD;
```

Comment:
Interesting we don't have a utility method to encapsulate "END_OF_TERM" already. Worth considering maybe?